File: Pharmaceutical_Company_Example/Behavioral/Java/Command.java
// Command Interface
interface RegulatoryCommand {
    void execute();
}

// Concrete Command: Submit Drug Application
class SubmitApplicationCommand implements RegulatoryCommand {
    private String drugName;
    private RegulatorySystem system;

    public SubmitApplicationCommand(String drugName, RegulatorySystem system) {
        this.drugName = drugName;
        this.system = system;
    }

    @Override
    public void execute() {
        system.submitApplication(drugName);
    }
}

// Concrete Command: Approve Drug
class ApproveDrugCommand implements RegulatoryCommand {
    private String drugName;
    private RegulatorySystem system;

    public ApproveDrugCommand(String drugName, RegulatorySystem system) {
        this.drugName = drugName;
        this.system = system;
    }

    @Override
    public void execute() {
        system.approveDrug(drugName);
    }
}

// Concrete Command: Reject Drug
class RejectDrugCommand implements RegulatoryCommand {
    private String drugName;
    private RegulatorySystem system;

    public RejectDrugCommand(String drugName, RegulatorySystem system) {
        this.drugName = drugName;
        this.system = system;
    }

    @Override
    public void execute() {
        system.rejectDrug(drugName);
    }
}

// Receiver
class RegulatorySystem {
    private java.util.Map<String, Boolean> applications;
    private java.util.Map<String, String> status;

    public RegulatorySystem() {
        applications = new java.util.HashMap<>();
        status = new java.util.HashMap<>();
    }

    public void submitApplication(String drugName) {
        applications.put(drugName, true);
        status.put(drugName, "Submitted");
        System.out.println("Application submitted for: " + drugName);
    }

    public void approveDrug(String drugName) {
        if (applications.containsKey(drugName)) {
            status.put(drugName, "Approved");
            System.out.println(drugName + " approved.");
        } else {
            System.out.println("No application found for: " + drugName);
        }
    }

    public void rejectDrug(String drugName) {
        if (applications.containsKey(drugName)) {
            status.put(drugName, "Rejected");
            System.out.println(drugName + " rejected.");
        } else {
            System.out.println("No application found for: " + drugName);
        }
    }

    public void showStatus(String drugName) {
        if (status.containsKey(drugName)) {
            System.out.println("Current status of " + drugName + ": " + status.get(drugName));
        } else {
            System.out.println("No status found for: " + drugName);
        }
    }
}

// Invoker
class ApprovalSystemInvoker {
    private java.util.List<RegulatoryCommand> commands;

    public ApprovalSystemInvoker() {
        commands = new java.util.ArrayList<>();
    }

    public void addCommand(RegulatoryCommand command) {
        commands.add(command);
    }

    public void executeCommands() {
        for (RegulatoryCommand command : commands) {
            command.execute();
        }
        commands.clear(); // Clear the command queue after execution
    }
}

public class Main {
    public static void main(String[] args) {
        // Create Receiver
        RegulatorySystem regulatorySystem = new RegulatorySystem();

        // Create Invoker
        ApprovalSystemInvoker approvalSystemInvoker = new ApprovalSystemInvoker();

        // Add commands to invoker
        SubmitApplicationCommand submitVaccineCommand = new SubmitApplicationCommand("Test Small Molecule 01", regulatorySystem);
        ApproveDrugCommand approveVaccineCommand = new ApproveDrugCommand("Test Small Molecule 01", regulatorySystem);
        RejectDrugCommand rejectAspirinCommand = new RejectDrugCommand("Test Small Molecule 02", regulatorySystem);

        approvalSystemInvoker.addCommand(submitVaccineCommand);
        approvalSystemInvoker.addCommand(approveVaccineCommand);
        approvalSystemInvoker.addCommand(rejectAspirinCommand);

        // Execute all commands
        approvalSystemInvoker.executeCommands();

        // Show the status of drugs after executing commands
        regulatorySystem.showStatus("Test Small Molecule 01");
        regulatorySystem.showStatus("Test Small Molecule 02");
    }
}

File: Pharmaceutical_Company_Example/03_Behavioral/C++/Command.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <memory>

using namespace std;

// Command Interface
class RegulatoryCommand {
public:
  virtual void execute() = 0;
  virtual ~RegulatoryCommand() = default;
};

// Receiver
class RegulatorySystem {
private:
  unordered_map<string, string> applications;
  
public:
  void submitApplication(const string& drugName) {
    applications[drugName] = "Submitted";
    cout << "Application submitted for: " << drugName << endl;
  }
  
  void approveDrug(const string& drugName) {
    if (applications.find(drugName) != applications.end()) {
      applications[drugName] = "Approved";
      cout << drugName << " approved." << endl;
    } else {
      cout << "No application found for: " << drugName << endl;
    }
  }
  
  void rejectDrug(const string& drugName) {
    if (applications.find(drugName) != applications.end()) {
      applications[drugName] = "Rejected";
      cout << drugName << " rejected." << endl;
    } else {
      cout << "No application found for: " << drugName << endl;
    }
  }
  
  void showStatus(const string& drugName) const {
    auto it = applications.find(drugName);
    if (it != applications.end()) {
      cout << "Current status of " << drugName << ": " << it->second << endl;
    } else {
      cout << "No status found for: " << drugName << endl;
    }
  }
};

// Concrete Command: Submit Drug Application
class SubmitApplicationCommand : public RegulatoryCommand {
private:
  string drugName;
  RegulatorySystem& system;
  
public:
  SubmitApplicationCommand(const string& drugName, RegulatorySystem& system) 
    : drugName(drugName), system(system) {}
  
  void execute() override {
    system.submitApplication(drugName);
  }
};

// Concrete Command: Approve Drug
class ApproveDrugCommand : public RegulatoryCommand {
private:
  string drugName;
  RegulatorySystem& system;
  
public:
  ApproveDrugCommand(const string& drugName, RegulatorySystem& system) 
    : drugName(drugName), system(system) {}
  
  void execute() override {
    system.approveDrug(drugName);
  }
};

// Concrete Command: Reject Drug
class RejectDrugCommand : public RegulatoryCommand {
private:
  string drugName;
  RegulatorySystem& system;
  
public:
  RejectDrugCommand(const string& drugName, RegulatorySystem& system) 
    : drugName(drugName), system(system) {}
  
  void execute() override {
    system.rejectDrug(drugName);
  }
};

// Invoker
class ApprovalSystemInvoker {
private:
  vector<unique_ptr<RegulatoryCommand>> commands;
  
public:
  void addCommand(unique_ptr<RegulatoryCommand> command) {
    commands.push_back(move(command));
  }
  
  void executeCommands() {
    for (auto& command : commands) {
      command->execute();
    }
    commands.clear(); // Clear the command queue after execution
  }
};

int main() {
  // Create Receiver
  RegulatorySystem regulatorySystem;
  
  // Create Invoker
  ApprovalSystemInvoker approvalSystemInvoker;
  
  // Add commands to invoker
  approvalSystemInvoker.addCommand(make_unique<SubmitApplicationCommand>("Test Small Molecule 01", regulatorySystem));
  approvalSystemInvoker.addCommand(make_unique<ApproveDrugCommand>("Test Small Molecule 01", regulatorySystem));
  approvalSystemInvoker.addCommand(make_unique<RejectDrugCommand>("Test Small Molecule 02", regulatorySystem));
  
  // Execute commands
  approvalSystemInvoker.executeCommands();
  
  // Show the status of drugs after executing commands
  regulatorySystem.showStatus("Test Small Molecule 01");
  regulatorySystem.showStatus("Test Small Molecule 02");
  
  return 0;
}
